packages/fxa-payments-server/server/lib/routes/legal-docs.js

```diff
@@ -0,0 +1,184 @@
+const acceptLanuageParser = require('accept-language-parser');
+const supportedLanguages = require('fxa-shared/l10n/supportedLanguages.json');
+const DEFAULT_LOCALE = 'en';
+const CACHE_DURATION = 24 * 60 * 60; // a day in seconds
+const LANGUAGES_LIMIT = 3;
+
+/**
+ * A map for caching.  The key is a url and the value is an object with two properties:
+ *  - checkedOn: timestamp (in ms) when the last check against the url was performed
+ *  - result: boolean
+ *
+ * The size of this map will also be limited.  There are 71 supported languages
+ * and two docs per product (not price/plan).  With some generous rounding, we
+ * have 200 per product.  The current default is set to 65536, which should
+ * give us a decent runway.
+ */
+const cache = new Map();
+
+const getFullPath = (pathWithoutLocale, language) =>
+  // minimal handling of an errant '/'
+  pathWithoutLocale.endsWith('/')
+    ? `${pathWithoutLocale.substring(
+        0,
+        pathWithoutLocale.length - 1
+      )}.${language}.pdf`
+    : `${pathWithoutLocale}.${language}.pdf`;
+
+/**
+ * @param {obj} lang A language object from accept-language-parser
+ */
+const getLanguageTag = (lang) =>
+  lang.region ? `${lang.code}-${lang.region}` : lang.code;
+
+const isSupportedLanguage = (tag) => supportedLanguages.includes(tag);
+
+const isCacheExpired = (lastChecked) =>
+  Math.round((Date.now() - lastChecked) / 1000) > CACHE_DURATION;
+
+/**
+ * Keeping this simple: if the cache's full, delete half of the entries.
+ * It mutates the map from the arguments directly.
+ */
+const checkAndShrinkCache = (cacheMap, limit) => {
+  if (limit > 0 && cacheMap.size >= limit) {
+    const half = Math.ceil(limit / 2);
+    const keys = Array.from(cacheMap.keys()).slice(0, half);
+    for (const k of keys) {
+      cacheMap.delete(k);
+    }
+  }
+};
+
+const isValidDocUrl = async (url) => {
+  // Check cache first
+  const cachedVal = cache.get(url);
+  if (cachedVal !== undefined) {
+    if (isCacheExpired(cachedVal['checkedOn'])) {
+      cache.delete(url);
+    } else {
+      return cachedVal['result'];
+    }
+  }
+
+  checkAndShrinkCache(cache, resultsCacheLimit);
+
+  try {
+    const res = await fetch(url, { method: 'HEAD' });
+    const entry = { checkedOn: Date.now(), result: res.status === 200 };
+    cache.set(url, entry);
+    return entry['result'];
+  } catch {
+    // NOOP
+    // fetch API exceptions are not 3xx-5xx responses
+  }
+
+  return false;
+};
+
+module.exports = {
+  method: 'get',
+  path: '/legal-docs',
+  validate: {
+    query: { url: joi.string().uri().required() },
+  },
+  async process(req, res) {
+    const docUrl = new URL(req.query.url);
+
+    // Limit the redirect to just the CDN that's hosting the legal docs
+    if (docUrl.hostname !== cdnFqdn) {
+      return res.sendStatus(400).end();
+    }
+
+    // For backwards compatibility and future single-locale products(?)
+    if (docUrl.pathname.endsWith('.pdf')) {
```

Comment:
How does this compare to the new form of `docUrl`?

Reply:
The new pattern is from https://github.com/mozilla/fxa/blob/main/_scripts/legal-md-to-pdf.sh#L107

Reply:
That pattern is for a path relative to a local file system though, not a URL right? I guess I'm not understanding how the new pattern's pathname doesn't also end with `'.pdf'`. 🤔

Reply:
The PDFs generated by that script is deployed to the CDN.  (Take a look at the `assets` dir under the FxA project root dir.)

The url in the query params will be from the product metadata, so they can be... anything really.  But when properly configured, it should be a direct url to a PDF, or, one without the locale and file extension, relying on this endpoint to find the best-match locale.